### PR TITLE
[SPARK-37455][SQL] Replace hash with sort aggregate if child is already sorted

### PR DESCRIPTION
**Pull request by c21**
### What changes were proposed in this pull request?

In the query plan, if the child of hash aggregate is already sorted on group-by columns, we can replace hash aggregate with sort aggregate for better performance, as sort aggregate does not have hashing overhead of hash aggregate. Add a physical plan rule `ReplaceHashWithSortAgg` here, and can be disabled by config `spark.sql.execution.replaceHashWithSortAgg`.

In addition, to help review as this PR triggers several TPCDS plan files change. The files below are having the real code change:

* `SQLConf.scala`
* `QueryExecution.scala`
* `ReplaceHashWithSortAgg.scala`
* `AdaptiveSparkPlanExec.scala`
* `HashAggregateExec.scala`
* `ReplaceHashWithSortAggSuite.scala`
* `SQLMetricsSuite.scala`

### Why are the changes needed?

To get better query performance by leveraging sort ordering in query plan.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Added unit test in `ReplaceHashWithSortAggSuite.scala`.

### CODE REVIEW

**Comment by c21** [comment]
cc @cloud-fan could you help take a look when you have time? Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala [hunk 1]
```diff
@@ -423,6 +423,9 @@ object QueryExecution {
       PlanSubqueries(sparkSession),
       RemoveRedundantProjects,
       EnsureRequirements(),
+      // `ReplaceHashWithSortAgg` needs to be added after `EnsureRequirements` to guarantee the
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala:426]: Is it because the planner is top-down so we don't know the child ordering during planning? Then we have to add a new rule to change the agg algorithm in a post-hoc way.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ReplaceHashWithSortAgg.scala [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+  private def replaceHashAgg(plan: SparkPlan): SparkPlan = {
+    plan.transformDown {
+      case hashAgg: HashAggregateExec =>
+        val sortAgg = hashAgg.toSortAggregate
+        hashAgg.child match {
+          case partialAgg: HashAggregateExec if isPartialAgg(partialAgg, hashAgg) =>
+            if (SortOrder.orderingSatisfies(
+                partialAgg.child.outputOrdering, sortAgg.requiredChildOrdering.head)) {
+              sortAgg.copy(
+                aggregateExpressions = sortAgg.aggregateExpressions.map(_.copy(mode = Complete)),
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ReplaceHashWithSortAgg.scala:75]: is it always right? I think we also need to check the output partitioning to see if we can eliminate the partial agg.

An example is `df.sortWithinPartitions`. It does not cluster the data, just sort it within each partition.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala (hunk 1 above)
Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala:426]: Yes it is. If we change our planning to bottom-up and propagate each node output ordering info during planning, then we can run this rule during planning. For now, we have to add it after `EnsureRequirements`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ReplaceHashWithSortAgg.scala (hunk 2 above)
Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/ReplaceHashWithSortAgg.scala:75]: @cloud-fan - I don't think we need to check output partitioning, as we are matching a pair of final and partial hash agg, without shuffle in between:

```
  HashAggregate(final)
          |                                       SortAggregate(complete)
HashAggregate(partial)             =>                    |
          |                                            child
        child 
```

So `child` must already have proper output partitioning for `SortAggregate`, o.w. it cannot satisfy original `HashAggregate(final)`'s required distribution.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/ReplaceHashWithSortAgg.scala:75]: ah ok, if there is a shuffle in the middle, we can't optimize? This looks quite limited, as having a shuffle in the middle is very common.

Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/ReplaceHashWithSortAgg.scala:75]: > if there is a shuffle in the middle, we can't optimize?

We can, and the rule here also does pattern matching for single `HashAggregate` below. I added a unit test case in `ReplaceHashWithSortAggSuite.scala` to demonstrate replacing partial aggregate - `"replace partial hash aggregate with sort aggregate"`. But I think it would be rare to be able to replace final aggregate (though this rule also covers it), as final aggregate is almostly always immediately after a shuffle, so there's no sort ordering before final aggregate.

Spark native shuffle does not guarantee any sort orders, for Cosco (a remote shuffle service we are running in-house), we support sorted shuffle, so final aggregate can also be possible to replace.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 3]
```diff
@@ -694,7 +694,8 @@ class SQLMetricsSuite extends SharedSparkSession with SQLMetricsTestUtils
   }
 
   test("SPARK-25497: LIMIT within whole stage codegen should not consume all the inputs") {
-    withSQLConf(SQLConf.WHOLESTAGE_CODEGEN_ENABLED.key -> "true") {
+    withSQLConf(SQLConf.WHOLESTAGE_CODEGEN_ENABLED.key -> "true",
+      SQLConf.REPLACE_HASH_WITH_SORT_AGG_ENABLED.key -> "false") {
```
Comment (c21) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala:698]: This unit test is testing multiple limit operators and hash aggregate operators in one single stage. Disable the rule here because sort aggregate does not support code-gen now, and will break the test logic.

**Comment by c21** [comment]
This PR is ready for review again thanks @cloud-fan.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ReplaceHashWithSortAgg.scala [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+            }
+        }
+      case other => other
+    }
+  }
+
+  /**
+   * Check if `partialAgg` to be partial aggregate of `finalAgg`.
+   */
+  private def isPartialAgg(partialAgg: HashAggregateExec, finalAgg: HashAggregateExec): Boolean = {
```
Comment (tanelk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ReplaceHashWithSortAgg.scala:95]: This looks like reverse enginering the `AggUtils`. Could we just link the partial and final agg when they are constructed?

Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/ReplaceHashWithSortAgg.scala:95]: @tanelk - yeah I agree this is mostly reverse engineering and we can do a better job here. I tried link partial and final agg in `AggUtils` and check linked physical plan to be same or not. This does not quite work due to we are doing top-down planning, and the linked partial agg not being same as planned partial agg (having `PlanLater` operator in linked partial agg).

I found a more elegant way to do it, by checking the linked logical plan of both aggs to be same. Updated.

Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/ReplaceHashWithSortAgg.scala:95]: cc @cloud-fan for review, thanks.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ReplaceHashWithSortAgg.scala [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+ * or
+ * 2. The plan is a [[HashAggregateExec]], and the child satisfies the sort order of
+ *    corresponding [[SortAggregateExec]].
+ *
+ * Examples:
+ * 1. aggregate after join:
+ *
+ *  HashAggregate(t1.i, SUM, final)
+ *               |                         SortAggregate(t1.i, SUM, complete)
+ * HashAggregate(t1.i, SUM, partial)   =>                |
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ReplaceHashWithSortAgg.scala:40]: This seems like an orthogonal optimization: we can merge adjacent partial and final aggregates (no shuffle between them) into one complete aggregate.

**Comment by cloud-fan** [comment]
The code change LGTM. Since there are quite some TPCDS queries that get plan changes, can we run a TPCDS benchmark to verify performance improvement?

**Comment by c21** [comment]
> Since there are quite some TPCDS queries that get plan changes, can we run a TPCDS benchmark to verify performance improvement?

@cloud-fan - sure. Today I ran the TPCDS benchmark (sf=1) on one AWS `r3.xlarge` (same as https://github.com/apache/spark/pull/26049). I don't see much performance difference compared enabling and disabling this rule:

* [Benchmark result with enabling this rule](https://gist.github.com/c21/b253d2d0ab8091b2953a633152927204)
* [Benchmark result with disabling this rule](https://gist.github.com/c21/d16583e5acb35b402af25c47da45248a)

Then I tried with sf=5, but the benchmark has task failure with no space left on device, so the benchmark cannot be conducted on single machine.

Do you recommend disabling this rule by default? After adding sort aggregate code-gen, we can do more large scale testing to enable it. WDYT?

**Comment by c21** [comment]
Discussed offline with @cloud-fan, we decide to disable the rule by default now. After adding sort aggregate code-gen, a large-scale TPCDS benchmark can be done later.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ReplaceHashWithSortAgg.scala (hunk 5 above)
Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/ReplaceHashWithSortAgg.scala:40]: Yeah I think we can add a rule later to optimize it. I vaguely remember someone proposed this in OSS before but seems impact is not high.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by c21** [comment]
Thank you @cloud-fan and @tanelk for review!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ReplaceHashWithSortAgg.scala [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+      replaceHashAgg(plan)
+    }
+  }
+
+  /**
+   * Replace [[HashAggregateExec]] with [[SortAggregateExec]].
+   */
+  private def replaceHashAgg(plan: SparkPlan): SparkPlan = {
+    plan.transformDown {
+      case hashAgg: HashAggregateExec if hashAgg.groupingExpressions.nonEmpty =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ReplaceHashWithSortAgg.scala:68]: BTW, shall we handle `ObjectHashAggregateExec` as well?

Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/ReplaceHashWithSortAgg.scala:68]: @cloud-fan - yeah I agree. Don't see a problem why we cannot do it. Created https://issues.apache.org/jira/browse/SPARK-37557 for followup. Will do it shortly, thanks.
